### PR TITLE
[patch-agent-onton-integration] Patch 3: Pure event mapper: Patch_agent_rpc.event -> Stream_event.t

### PR DESCRIPTION
**Pull request by subsetpark**
## Patch 3: Pure event mapper: Patch_agent_rpc.event -> Stream_event.t

- Implement `val map_event : Patch_agent_rpc.event -> Types.Stream_event.t`. Mapping: Session_init -> Session_init { session_id }; Turn_started -> Turn_started; Text_delta -> Text_delta; Tool_call -> Tool_use { name; input; status = None } (patch-agent emits tool_call once per call; status is None per the existing convention for non-OpenCode backends); Done -> Final_result { text = final_text; stop_reason = parsed from string per Stop_reason.t }; Error -> Error.
- Define a small helper `val parse_stop_reason : string -> Stop_reason.t` mapping patch-agent's stop_reason strings ("stop", "tool_use", "max_tokens", "error") to onton's Stop_reason.t variants. Unknown strings map to a documented default and emit a warning event log entry.
- Property tests: mapping is total (no exception over fc-generated event values); session_id is preserved verbatim; tool_call name and input are preserved verbatim; final_result text is preserved verbatim; map_event is deterministic.

## Changes
- Implement `val map_event : Patch_agent_rpc.event -> Types.Stream_event.t`. Mapping: Session_init -> Session_init { session_id }; Turn_started -> Turn_started; Text_delta -> Text_delta; Tool_call -> Tool_use { name; input; status = None } (patch-agent emits tool_call once per call; status is None per the existing convention for non-OpenCode backends); Done -> Final_result { text = final_text; stop_reason = parsed from string per Stop_reason.t }; Error -> Error.
- Define a small helper `val parse_stop_reason : string -> Stop_reason.t` mapping patch-agent's stop_reason strings ("stop", "tool_use", "max_tokens", "error") to onton's Stop_reason.t variants. Unknown strings map to a documented default and emit a warning event log entry.
- Property tests: mapping is total (no exception over fc-generated event values); session_id is preserved verbatim; tool_call name and input are preserved verbatim; final_result text is preserved verbatim; map_event is deterministic.

## Files to Modify
- lib/patch_agent_event_mapper.ml (create): Total mapping function from patch-agent events to onton's canonical Stream_event.t.
- lib/patch_agent_event_mapper.mli (create): Public interface for the mapper.
- test/test_patch_agent_event_mapper.ml (create): QCheck2 property tests for totality and field preservation.
- lib/dune (modify): Add patch_agent_event_mapper to modules list.
- test/dune (modify): Add test_patch_agent_event_mapper executable.

## Gameplan Specification

```
module PATCH_AGENT_INTEGRATION.

> M4 ships a long-lived backend abstraction (Llm_backend_long_lived)
> distinct from the ephemeral fresh-spawn shape, an implementation
> (Patch_agent_backend) that drives the patch-agent stdio RPC, and
> kind-driven dispatch in Patch_controller. The patch-agent backend
> is registered under Backend_registry like every other backend;
> selection follows the existing --backend and modelRouting
> machinery. Existing ephemeral backends are untouched.

Backend.
BackendKind.
ephemeral => BackendKind.
long-lived => BackendKind.
patch-agent-backend => Backend.

Session.
LifecyclePhase.
start-phase => LifecyclePhase.
prompt-phase => LifecyclePhase.
abort-phase => LifecyclePhase.
shutdown-phase => LifecyclePhase.

RpcEvent.
StreamEvent.
map-event e: RpcEvent => StreamEvent.

kind-of b: Backend => BackendKind.
long-lived-dispatch? b: Backend => Bool.

start-count s: Session => Nat0.
shutdown-count s: Session => Nat0.
prompt-after-shutdown? s: Session => Bool.
---
> Lifecycle: each long-lived Session has exactly one start, at most
> one shutdown, and no prompt or abort after shutdown.
all s: Session | start-count s = 1.
all s: Session | shutdown-count s <= 1.
all s: Session | ~prompt-after-shutdown? s.

> The patch-agent backend has kind long-lived.
kind-of patch-agent-backend = long-lived.

> Long-lived dispatch is used iff the backend has kind long-lived.
all b: Backend | long-lived-dispatch? b <-> kind-of b = long-lived.

```

## Patch Specification

```
module PATCH_AGENT_INTEGRATION_PATCH_3.

> Patch 3 introduces the event mapper. The function signature
> `map-event e: RpcEvent => StreamEvent` is itself the totality
> guarantee: the mapper is defined for every RpcEvent.

RpcEvent.
StreamEvent.
map-event e: RpcEvent => StreamEvent.
---
true.

```

## Implementation Notes

- `parse_stop_reason` maps `"stop"` to `End_turn`, `"tool_use"` to `Tool_use`, and `"max_tokens"` to `Max_tokens`. Patch-agent's `"error"` string does not have a direct `Types.Stop_reason.t` counterpart in onton, so it falls back to `End_turn`; the same fallback is used for any other unknown string and is documented in the `.mli`.
- The original note about emitting an event-log warning could not be implemented literally in this patch without making the mapper effectful or threading an `Event_log.t`/warning callback through the API. The mapper stays total with the requested `string -> Stop_reason.t` signature and emits a single stderr warning instead.
- `Patch_agent_rpc.Error` carries both `code` and `message`; `Stream_event.Error` only carries one string, so the mapper preserves both by formatting `"patch-agent <code>: <message>"` when a code is present.
- `lib/dune` did not need a manual modules-list edit in this repo layout; dune picks up the new library module automatically. The only dune wiring change was the new test stanza in `test/dune`.

---
## Summary by cubic
Adds a pure, total mapper from `Patch_agent_rpc.event` to `Types.Stream_event.t` to standardize patch-agent output. Includes stop-reason parsing (stderr warning for unknown values only) and property tests for totality, determinism, field preservation, and stop-reason mapping.

- New Features
  - Added `map_event` covering all RPC variants; maps Tool_call -> Tool_use (status=None), Done -> Final_result with parsed stop_reason, and formats Error as "patch-agent <code>: <message>" when a code is present; whitespace-only codes use the message.
  - Added `parse_stop_reason` supporting "stop", "tool_use", "max_tokens"; defers to `Types.Stop_reason.of_string` when possible; "error" defaults to End_turn silently; unknown values default to End_turn and emit a stderr warning.

<sup>Written for commit 44f9f21cfd6d4ad2ff07ceba1f8c7da3b858b892. Summary will update on new commits.</sup>

## Summary by CodeRabbit

* **New Features**
  * Converts incoming RPC events into the app's standardized stream events, normalizes stop reasons with a safe fallback and warning for unknown values, and cleans up error messages by stripping embedded codes.

* **Tests**
  * Added property-based tests to validate event mapping correctness, determinism, preservation of key fields, stop-reason handling, and edge-case behaviors.

### CODE REVIEW

File: lib/patch_agent_rpc.ml [hunk 1]
```diff
@@ -81,14 +81,21 @@ let require_int field json =
 let parse_usage json =
   let ( let* ) = Result.( >>= ) in
   match json with
+  | `Null -> Result.Ok None
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_agent_rpc.ml]: Addressed in later commits during the rebase onto the Patch 2 base. The RPC codec now lives in `lib_core/patch_agent_rpc.ml`, `parse_usage` rejects `"usage": null` when the field is present, and `test/test_patch_agent_rpc.ml` includes a regression test for that case.

File: lib/patch_agent_event_mapper.ml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      (Printf.sprintf
+         "warning: patch-agent reported unsupported stop_reason %S; defaulting \
+          to end_turn"
+         raw))
+
+let parse_stop_reason (raw : string) : Types.Stop_reason.t =
+  match raw with
+  | "stop" -> Types.Stop_reason.End_turn
+  | "tool_use" -> Types.Stop_reason.Tool_use
+  | "max_tokens" -> Types.Stop_reason.Max_tokens
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_agent_event_mapper.ml:14]: Fixed in `8eeb5f5`. The known patch-agent stop reason `"error"` now maps directly to `End_turn` without going through the unsupported-value warning path, so we only warn for truly unknown/unparseable strings.

File: lib/patch_agent_event_mapper.ml [hunk 3]
```diff
@@ -0,0 +1,47 @@
+open Base
+
+let warned_unknown_stop_reason = ref false
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_agent_event_mapper.ml]: Fixed in `8eeb5f5`. I removed the process-global `bool ref` dedupe and now emit the warning on every unsupported stop-reason value, so `parse_stop_reason` no longer carries hidden cross-test/process state.

File: test/test_patch_agent_event_mapper.ml [hunk 4]
```diff
@@ -0,0 +1,113 @@
+open Base
+open Onton
+open Onton_core
+
+let gen_char =
+  QCheck2.Gen.map
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_patch_agent_event_mapper.ml:6]: Fixed in `8eeb5f5`. I added a targeted property for `Patch_agent_rpc.Error { code = "   "; message = "oops" }`, which exercises the `String.strip code = ""` branch and asserts that the mapper emits `Types.Stream_event.Error "oops"`.

File: test/test_patch_agent_event_mapper.ml [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+let () =
+  let suite =
+    [
+      prop_total;
+      prop_session_id_preserved;
+      prop_tool_call_preserved;
+      prop_final_text_preserved;
+      prop_whitespace_code_uses_message;
+      prop_deterministic;
+    ]
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply test/test_patch_agent_event_mapper.ml:138]: Added in `9975262`. I introduced a table-style contract test that exercises `Patch_agent_rpc.Done` through `Patch_agent_event_mapper.map_event` for `"stop"`, `"tool_use"`, `"max_tokens"`, `"error"`, and an unknown future value, and asserts the expected `Types.Stream_event.Final_result` mapping for each.

File: lib/patch_agent_event_mapper.mli [hunk 6]
```diff
@@ -0,0 +1,13 @@
+open Base
+
+(** Total mapping from patch-agent stdio RPC events into onton's canonical
+    [Types.Stream_event.t].
+
+    Patch-agent exposes stop-reason strings that do not exactly match onton's
+    variant names. ["stop"] maps to [Types.Stop_reason.End_turn]. Unsupported
+    values, including patch-agent's ["error"], also default to
+    [Types.Stop_reason.End_turn]. The mapper emits a warning to stderr when it
```
[the remark this answers is not in the document]
Comment (subsetpark) [reply lib/patch_agent_event_mapper.mli]: Updated the interface docstring to match the current implementation: patch-agent's "error" stop_reason is an explicit silent mapping to End_turn, while only fully unknown stop_reason strings warn on fallback. Verified with dune build and dune runtest.
